theano/contrib/mkl/mkl_ndarray.c

```diff
@@ -0,0 +1,1361 @@
+    0,                         /*tp_setattr*/
+    0,                         /*tp_compare*/
+    MKLNdarray_repr,           /*tp_repr*/
+    0,                         /*tp_as_number*/
+    0,                         /*tp_as_sequence*/
+    0,                         /*tp_as_mapping*/
+    0,                         /*tp_hash */
+    0,                         /*tp_call*/
+    0,                         /*tp_str*/
+    0,                         /*tp_getattro*/
+    0,                         /*tp_setattro*/
+    0,                         /*tp_as_buffer*/
+#if PY_MAJOR_VERSION >= 3
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
+#else
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_CHECKTYPES, /*tp_flags*/
+#endif
+    "MKLNdarray objects",      /*tp_doc */
+    0,                         /*tp_traverse*/
+    0,                         /*tp_clear*/
+    0,                         /*tp_richcompare*/
+    0,                         /*tp_weaklistoffset*/
+    0,                         /*tp_iter*/
+    0,                         /*tp_iternext*/
+    MKLNdarray_methods,        /*tp_methods*/
+    MKLNdarray_members,        /*tp_members*/
+    MKLNdarray_getset,         /*tp_getset*/
+    0,                         /*tp_base*/
+    0,                         /*tp_dict*/
+    0,                         /*tp_descr_get*/
+    0,                         /*tp_descr_set*/
+    0,                         /*tp_dictoffset*/
+    (initproc)MKLNdarray_init, /*tp_init*/
+    0,                         /*tp_alloc*/
+    MKLNdarray_new,            /*tp_new*/
+};
+
+
+/*
+ * Check an input is an instance of MKLNdarray or not.
+ * Same as PyArray_Check
+ *
+ */
+int MKLNdarray_Check(const PyObject *ob) {
+    return ((Py_TYPE(ob) == &MKLNdarrayType) ? 1 : 0);
+}
+
+
+/*
+ * Try to new a MKLNdarray instance.
+ * This function is different from MKLNdarray_new which is set for tp_new.
+ *
+ * MKLNdarray_new is call by python when python has allocate memory for it already.
+ * MKLNdarray_New will be called manually in C/C++ code, so we need to call tp_alloc manually.
+ *
+ */
+PyObject*
+MKLNdarray_New(int nd, int typenum) {
+    if (nd < 0 || nd > MAX_NDIM) {
+        PyErr_Format(PyExc_ValueError,
+                     "MKLNdarray_New: not support a %d-dim array. Try array which ndim is <= %d. line: %d",
+                     nd, MAX_NDIM, __LINE__);
+        return NULL;
+    }
+
+    MKLNdarray* self = (MKLNdarray*)(MKLNdarrayType.tp_alloc(&MKLNdarrayType, 0));
+    if (self == NULL) {
+        PyErr_SetString(PyExc_RuntimeError, "MKLNdarray_New: failed to allocate self");
+        return NULL;
+    }
+    self->base              = NULL;
+    self->nd                = nd;
+    self->dtype             = typenum;
+    self->private_data      = NULL;
+    self->private_workspace = NULL;
+    self->data_size         = 0;
+    self->private_layout    = NULL;
+    memset((void*)(self->user_structure), 0, 2 * MAX_NDIM * sizeof (size_t));
```

Comment:
Since we need strides information to build layout_user, so we need to keep this.